### ISSUE TITLE
Object spread incorrectly introduces undefined value type in TS 4.2

### ISSUE THREAD

**Issue by danvk** [microsoft/TypeScript#43045]
# Bug Report

This was the one type error I got when I updated my project from TS 4.1 → 4.2. It's a false positive so I figured I'd file an issue in case others run into it as well. I assume this is considered an acceptable loss from [this change](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2/#relaxed-optional-index-rules) in the release notes, but it is a false positive and regression from 4.1 nonetheless!

### 🔎 Search Terms

- object spread undefined 4.2 

Possibly related to #13195

### 🕗 Version & Regression Information

- This changed between versions 4.1.5 and 4.2.2

### ⏯ Playground Link

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```
</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=4.2.2#code/MYewdgzgLgBADgQwE4ILYCYYF4YFkFQAWAdCmACYioAUAlDADwwAMxArDAPwwDkAbggA2AVwCm6HjABcMMMMGCA3AFgAUGtCRYIAEYArbDADeamGfjI0ARhn8hYqzwA0p88XfVEKDF2Ne06AC+0saBtGqBKuqqmtAwAI5iSACehgDy+qLAUMSiYFBIAJaiENS6erTEqAhw1K5m1ADaANZOMHwAuvRYAHwwAAYAJEbNgVjDeaDkogCqAEoAkgDCVHDgeVDUfLSB-fUwAPQH5ienZ+cXJwB++0cwAIJIAObCqBswIABmMFDJcKK8aBFMBPGAAHxgwgook+hTAonIkkKEFkIFgCAgEEKTzACB0ggBUBAFm8oigoiQH2+v3+gIKcNBELkqB0FPBMB0IBABIQYB4xFuxzMABU-gCeFDprD4YiYMjUejMdjcfjCcSaeKgQz2czWZSIZzuaJefzqOgAMwAFjY4VUlT0IDh1B4ADIeLQokA)

### 💻 Code

```ts
const param2 = Math.random() < 0.5 ? 'value2' : null;

const obj = {
    param1: 'value1',
    ...(param2 ? {param2} : {})
};

const query = Object.entries(obj).map(
    ([k, v]) => `${k}=${encodeURIComponent(v)}`
    //                                     ~
    // Argument of type 'string | undefined' is not assignable to parameter of type 'string | number | boolean'.
    //   Type 'undefined' is not assignable to type 'string | number | boolean'.(2345)
).join('&');
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const param2 = Math.random() < 0.5 ? 'value2' : null;
const obj = Object.assign({ param1: 'value1' }, (param2 ? { param2 } : {}));
const query = Object.entries(obj).map(([k, v]) => `${k}=${encodeURIComponent(v)}`
//                                     ~
// Argument of type 'string | undefined' is not assignable to parameter of type 'string | number | boolean'.
//   Type 'undefined' is not assignable to type 'string | number | boolean'.(2345)
).join('&');

```
</details>

### 🙁 Actual behavior

TypeScript says that `v` can be `undefined`, which it cannot. This leads to the type error in the call to `encodeURIComponent`.

### 🙂 Expected behavior

`v` can never be `undefined` at runtime, so the error is a false positive. This was not an error in TypeScript 4.1.

The root cause is that the type of `obj` is inferred as

```ts
const obj: {
    param2?: string | undefined;
    param1: string;
}
```

whereas in TS 4.1 the same inferred type shows up as:

```ts
const obj: {
    param2?: string;
    param1: string;
}
```

**Comment by dwelle** [comment]
This also applies to spreading nullable objects:

```
const obj = {} as ({value: string} | null);
const config = { ...obj }; //  { value?: string | undefined }
```

```
const obj = {} as ({value: string} | undefined);
const config = { ...obj }; //  { value?: string | undefined }
```

[playground](https://www.typescriptlang.org/play?ts=4.2.2#code/N4WAUABFEMYPYDsDOAXCcBGArCBeCwAvhAIZIQAUwAbiQDYCuApgFwSoBOAlggObEAfCAgZ06ASgDc4aLESo5CAGZdeeAgDotmLIUkQA9AajAItRkwD8bTjzVCGCACZMVCJk4iFw3sOFCQ0PDIaDrqRKTkVObMNijcfIIQji5uHlIyQfJowSpq+KZaGmF6hsYEZvTM1uzxdhAOzq48Hl4+4EA)

**Comment by ahejlsberg** [comment]
The change in 4.2 to add `undefined` to the type of optional properties created by spread expressions is definitely correct. An optional property should always have `undefined` in its type.

The real issue is more likely that we should remove `undefined` from the type of optional properties when inferring from the properties of some object type to an index signature (as is the case in the `Object.entries` method).

**Comment by dwelle** [comment]
@ahejlsberg unclear whether the fix for "inferring to index signatures" covers e.g. this use case (React's `setState`):

```ts
type State = { a: number };
const setState = <K extends keyof State>(data: Pick<State, K>) => {};

setState({ ...({} as { a: 42 } | null) }); // error in 4.2.2
```

[playground](https://www.typescriptlang.org/play?ts=4.2.2#code/C4TwDgpgBAysCGxoF4oG8rwFxQHYFcBbAIwgCcoBfAbgFgAoAYwHtcBnYKNiYORFKAB4A0lAgAPJLgAmbKAGsIIZgDNYCJAD4AFNMTYoABQCWjeYL5IANFGGaAlFGSb0NBg268NEbRgB0Ab6UmHIYBgAsAExUUAA+ePgANomOlPbUUAD0mWJkZMwUxrhQ4X6RZUA)

**Comment by ahejlsberg** [comment]
@dwelle Unfortunately no. To cover that scenario need the `missing` type such that optional properties can include `missing`, but not `undefined`, in their type.

**Comment by dwelle** [comment]
My bad, I thought TS < 4.2 allowed this, but it didn't. It just seems in our codebase we've used a pattern that was using some weird loophole that was now fixed in 4.2:

```ts
type State = { a: number };
const setState = <K extends keyof State>(updater: (state: State) => Pick<State, K> | State | null) => {};

const obj = {} as { a: 42 } | null | undefined; // <-- both `null | undefined` is needed
setState((state) => ({ ...obj })); // works in 4.1.5
```

Works in [4.1.5](https://www.typescriptlang.org/play?ts=4.1.5#code/C4TwDgpgBAysCGxoF4oG8rwFxQHYFcBbAIwgCcoBfAbgFgAoAYwHtcBnYKNiYORFKAB4A0lAgAPJLgAmbKAGsIIZgDNYCJAD4AFPjDT+ZHNo78cfJAEooyTVAAKAS0bzBFiABoowuwB91-FD+BAA2Ida26DQMDCzsnMzEAFY2UZhyGNhQACwATFRBePhhhfgyECqOuBDS1FAA9PVCALTNUMTMwAAWUAAGoSGl5ZXV0r1QjnLVNTUM3LwaENomixF22hgAdNuJKZSWltRAA), but not in [4.2.2](https://www.typescriptlang.org/play?ts=4.2.2#code/C4TwDgpgBAysCGxoF4oG8rwFxQHYFcBbAIwgCcoBfAbgFgAoAYwHtcBnYKNiYORFKAB4A0lAgAPJLgAmbKAGsIIZgDNYCJAD4AFPjDT+ZHNo78cfJAEooyTVAAKAS0bzBFiABoowuwB91-FD+BAA2Ida26DQMDCzsnMzEAFY2UZhyGNhQACwATFRBePhhhfgyECqOuBDS1FAA9PVCALTNUMTMwAAWUAAGoSGl5ZXV0r1QjnLVNTUM3LwaENomixF22hgAdNuJKZSWltRAA)

But, seeing as it's working as intended, there's no reason to go back to the old "buggy" behavior I guess.

Any rough ETA on #13195?
